### PR TITLE
Add SignalWithStart

### PR DESCRIPTION
**Pull request by vancexu**
releated https://github.com/uber/cadence/pull/621

### CODE REVIEW

File: .gen/go/shared/types.go [hunk 1]
```diff
@@ -21524,6 +21524,498 @@ func (v *SignalExternalWorkflowExecutionInitiatedEventAttributes) GetChildWorkfl
...
+// zero value if it is unset.
+func (v *SignalWithStartWorkflowExecutionRequest) GetRequestId() (o string) {
+	if v.RequestId != nil {
+		return *v.RequestId
+	}
+
+	return
+}
+
+// GetWorkflowIdReusePolicy returns the value of WorkflowIdReusePolicy if it is set or its
```
Comment (purplesusan) [inline .gen/go/shared/types.go]: GetWorkflowIdReusePolicy returns the value of WorkflowIdReusePolicy if it is set or its zero value if it is unset. ->
GetWorkflowIdReusePolicy returns the value of WorkflowIdReusePolicy if it is set, or a zero value if it is not set.

File: .gen/go/shared/types.go [hunk 2]
```diff
@@ -21524,6 +21524,498 @@ func (v *SignalExternalWorkflowExecutionInitiatedEventAttributes) GetChildWorkfl
...
+// zero value if it is unset.
+func (v *SignalWithStartWorkflowExecutionRequest) GetWorkflowIdReusePolicy() (o WorkflowIdReusePolicy) {
+	if v.WorkflowIdReusePolicy != nil {
+		return *v.WorkflowIdReusePolicy
+	}
+
+	return
+}
+
+// GetSignalName returns the value of SignalName if it is set or its
```
Comment (purplesusan) [inline .gen/go/shared/types.go]: GetSignalName returns the value of SignalName if it is set or its zero value if it is unset. ->
GetSignalName returns the value of SignalName if it is set, or a zero value if it is not set.

File: client/client.go [hunk 3]
```diff
@@ -104,6 +104,17 @@ type (
 		//	- InternalServiceError
 		SignalWorkflow(ctx context.Context, workflowID string, runID string, signalName string, arg interface{}) error
 
+		// SignalWithStartWorkflow sends a signals to a workflow in execution, if workflow is not running or not found,
+		// it starts workflow then sends the signal in transaction.
```
Comment (purplesusan) [inline client/client.go]: suggested update:
SignalWithStartWorkflow sends a signal to a running workflow. If the workflow is not running or not found, it starts the workflow and then sends the signal in transaction.

File: client/client.go [hunk 4]
```diff
@@ -104,6 +104,17 @@ type (
...
+		// - workflowID, runID, signalName, signalArg are same as SignalWorkflow's parameters
```
Comment (purplesusan) [inline client/client.go]: - workflowID, runID, signalName, signalArg are same as SignalWorkflow's parameters
The workflowID, runID, signalName, and signalArg parameters are the same as those of SignalWorkflow.

File: client/client.go [hunk 5]
```diff
@@ -104,6 +104,17 @@ type (
...
+		// - options, workflow, workflowArgs are same as StartWorkflow's parameters
```
Comment (purplesusan) [inline client/client.go]: - options, workflow, workflowArgs are same as StartWorkflow's parameters ->
The workflow and workflowArgs parameters are the same as those of StartWorkflow.

File: internal/client.go [hunk 6]
```diff
@@ -94,6 +94,17 @@ type (
...
 		SignalWorkflow(ctx context.Context, workflowID string, runID string, signalName string, arg interface{}) error
 
+		// SignalWithStartWorkflow sends a signals to a workflow in execution, if workflow is not running or not found,
+		// it starts workflow then sends the signal in transaction.
+		// - workflowID, runID, signalName, signalArg are same as SignalWorkflow's parameters
+		// - options, workflow, workflowArgs are same as StartWorkflow's parameters
+		// The errors it can return:
+		//  - EntityNotExistsError, if domain does not exists
+		//  - BadRequestError
+		//	- InternalServiceError
```
Comment (mfateev) [inline internal/client.go]: What's the semantic of the call when RunID is specified? For example previous workflow with this RunID has completed? Are we going to start a workflow ignoring RunID? It looks weird to me. I think we want to eliminate RunID parameter from this call as well as from the thrift API.

File: client/client.go [hunk 7]
```diff
@@ -104,6 +104,17 @@ type (
...
+		// The errors it can return:
+		//  - EntityNotExistsError, if domain does not exists
```
Comment (purplesusan) [inline client/client.go]: suggested update:
SignalWithStartWorkflow can return the following errors:
- EntityNotExistsError, if the domain does not exist

File: idl/github.com/uber/cadence/cadence.thrift [hunk 8]
```diff
@@ -298,6 +298,21 @@ service WorkflowService {
       4: shared.ServiceBusyError serviceBusyError,
     )
 
+  /**
+  * SignalWithStartWorkflowExecution is used to ensure sending a signal event to a workflow execution.
```
Comment (purplesusan) [inline idl/github.com/uber/cadence/cadence.thrift]: SignalWithStartWorkflowExecution is used to ensure sending a signal event to a workflow execution. ->
SignalWithStartWorkflowExecution is used to send a signal event to an executed workflow.

File: idl/github.com/uber/cadence/cadence.thrift [hunk 9]
```diff
@@ -298,6 +298,21 @@ service WorkflowService {
...
+  * If workflow is running, this results in WorkflowExecutionSignaled event recorded in the history
```
Comment (purplesusan) [inline idl/github.com/uber/cadence/cadence.thrift]: this results in WorkflowExecutionSignaled event recorded in the history -> this results in a WorkflowExecutionSignaled event being recorded in the history

File: idl/github.com/uber/cadence/cadence.thrift [hunk 10]
```diff
@@ -298,6 +298,21 @@ service WorkflowService {
...
+  * and a decision task being created for the execution.
+  * If workflow is not running or not found, this results in WorkflowExecutionStarted and WorkflowExecutionSignaled
```
Comment (purplesusan) [inline idl/github.com/uber/cadence/cadence.thrift]: If workflow is not running -> If the workflow is not running

File: idl/github.com/uber/cadence/cadence.thrift [hunk 11]
```diff
@@ -298,6 +298,21 @@ service WorkflowService {
...
+  * event recorded in history, and a decision task being created for the execution
```
Comment (purplesusan) [inline idl/github.com/uber/cadence/cadence.thrift]: event recorded in history, and a decision task being created for the execution -> events recorded in the history, and a decision task being created for the execution.

File: internal/client.go [hunk 12]
```diff
@@ -94,6 +94,17 @@ type (
 		//	- InternalServiceError
 		SignalWorkflow(ctx context.Context, workflowID string, runID string, signalName string, arg interface{}) error
 
+		// SignalWithStartWorkflow sends a signals to a workflow in execution, if workflow is not running or not found,
+		// it starts workflow then sends the signal in transaction.
```
Comment (purplesusan) [inline internal/client.go]: suggested update:
SignalWithStartWorkflow sends a signal to a running workflow. If the workflow is not running or not found, it starts the workflow and then sends the signal in transaction.

File: internal/client.go [hunk 13]
```diff
@@ -94,6 +94,17 @@ type (
...
+		// - workflowID, runID, signalName, signalArg are same as SignalWorkflow's parameters
```
Comment (purplesusan) [inline internal/client.go]: suggested update:
The workflowID, runID, signalName, and signalArg parameters are the same as those of SignalWorkflow.

File: internal/client.go [hunk 14]
```diff
@@ -94,6 +94,17 @@ type (
...
+		// - options, workflow, workflowArgs are same as StartWorkflow's parameters
```
Comment (purplesusan) [inline internal/client.go]: suggested update:
The options, workflow, and workflowArgs parameters are same as those of StartWorkflow.

File: internal/client.go [hunk 15]
```diff
@@ -94,6 +94,17 @@ type (
...
+		// The errors it can return:
+		//  - EntityNotExistsError, if domain does not exists
```
Comment (purplesusan) [inline internal/client.go]: suggested update:
SignalWithStartWorkflow can return the following errors:
- EntityNotExistsError, of the domain does not exist

File: internal/client.go [hunk 16]
```diff
@@ -94,6 +94,17 @@ type (
...
+		//  - BadRequestError
+		//	- InternalServiceError
+		SignalWithStartWorkflow(ctx context.Context, workflowID string, runID string, signalName string, signalArg interface{},
+			options StartWorkflowOptions, workflow interface{}, workflowArgs ...interface{}) (*WorkflowExecution, error)
+
 		// CancelWorkflow cancels a workflow in execution
```
Comment (purplesusan) [inline internal/client.go]: CancelWorkflow cancels a workflow in execution -> CancelWorkflow cancels a running workflow.

File: internal/internal_workflow_client.go [hunk 17]
```diff
@@ -263,6 +263,89 @@ func (wc *workflowClient) SignalWorkflow(ctx context.Context, workflowID string,
 		}, serviceOperationRetryPolicy, isServiceTransientError)
 }
 
+// SignalWithStartWorkflow sends a signals to a workflow in execution, if workflow is not running or not found,
```
Comment (purplesusan) [inline internal/internal_workflow_client.go]: suggested update:
SignalWithStartWorkflow sends a signal to a running workflow. If the workflow is not running or not found, it starts the workflow and then sends the signal in transaction.

File: internal/internal_workflow_client.go [hunk 18]
```diff
@@ -263,6 +263,89 @@ func (wc *workflowClient) SignalWorkflow(ctx context.Context, workflowID string,
...
+
+	decisionTaskTimeout := common.Int32Ceil(options.DecisionTaskStartToCloseTimeout.Seconds())
+	if decisionTaskTimeout < 0 {
+		return nil, errors.New("negative DecisionTaskStartToCloseTimeout provided")
+	}
+	if decisionTaskTimeout == 0 {
+		decisionTaskTimeout = defaultDecisionTaskTimeoutInSecs
+	}
+
+	// Validate type and its arguments.
```
Comment (purplesusan) [inline internal/internal_workflow_client.go]: Validate type and its arguments. -> Validate the workflow type and its arguments.

File: internal/internal_workflow_client.go [hunk 19]
```diff
@@ -263,6 +263,89 @@ func (wc *workflowClient) SignalWorkflow(ctx context.Context, workflowID string,
...
+		ExecutionStartToCloseTimeoutSeconds: common.Int32Ptr(executionTimeout),
+		TaskStartToCloseTimeoutSeconds:      common.Int32Ptr(decisionTaskTimeout),
+		SignalName:                          common.StringPtr(signalName),
+		SignalInput:                         signalInput,
+		Identity:                            common.StringPtr(wc.identity),
+	}
+
+	var response *s.StartWorkflowExecutionResponse
+
+	// Start creating workflow request.
```
Comment (purplesusan) [inline internal/internal_workflow_client.go]: Start creating workflow request. -> Create the workflow request.

File: internal/internal_workflow_client_test.go [hunk 20]
```diff
@@ -503,3 +503,81 @@ func getGetWorkflowExecutionHistoryRequest(filterType shared.HistoryEventFilterT
 
 	return request
 }
+
+// workflow client test suite
```
Comment (purplesusan) [inline internal/internal_workflow_client_test.go]: workflow client test suite -> Workflow client test suite.

File: mocks/Client.go [hunk 21]
```diff
@@ -286,6 +286,35 @@ func (_m *Client) SignalWorkflow(ctx context.Context, workflowID string, runID s
 	return r0
 }
 
+// SignalWithStartWorkflow provides a mock function with given fields: ctx, options, workflow, args
```
Comment (purplesusan) [inline mocks/Client.go]: SignalWithStartWorkflow provides a mock function with given fields: ctx, options, workflow, args ->
SignalWithStartWorkflow provides a mock function with the following fields: ctx, options, workflow, args.

File: internal/client.go (hunk 6 above)
Comment (vancexu) [reply internal/client.go]: Removed RunID. Agree having runID here might confuse user. User now only use workflowID to find latest running workflow to signal with start.

File: .gen/go/cadence/idl.go [hunk 22]
```diff
@@ -33,11 +33,11 @@ var ThriftModule = &thriftreflect.ThriftModule{
...
...rkflowResponse QueryWorkflow(1: shared.QueryWorkflowRequest queryRequest)\n\tthrows (\n\t  1: shared.BadRequestError badRequestError,\n\t  2: shared.InternalServiceError internalServiceError,\n\t  3: shared.EntityNotExistsError entityNotExistError,\n\t  4: shared.QueryFailedError queryFailedError,\n\t)\n\n  /**\n  * DescribeWorkflowExecution returns information about the specified workflow execution.\n  **/\n  shared.DescribeWorkflowExecutionResponse DescribeWorkflowExecution(1: shared.DescribeWorkflowExecutionRequest describeRequest)\n    throws (\n      1: shared.BadRequestError badRequestError,\n      2: shared.InternalServiceError internalServiceError,\n      3: shared.EntityNotExistsError entityNotExistError,\n    )\n\n  /**\n  * DescribeTaskList returns information about the target tasklist, right now this API returns the\n  * pollers which polled this tasklist in last few minutes.\n  **/\n  shared.DescribeTaskListResponse DescribeTaskList(1: shared.DescribeTaskListRequest request)\n    throws (\n      1: shared.BadRequestError badRequestError,\n      2: shared.InternalServiceError internalServiceError,\n      3: shared.EntityNotExistsError entityNotExistError,\n    )\n\n}\n"
```
[the remark this answers is not in the document]
Comment (vancexu) [reply .gen/go/cadence/idl.go]: dev team are advised to keep it 2017 from legal team.

**Comment by vancexu** [comment]
Thanks @purplesusan for the suggestions. files under ./gen is auto generated by tools, so we do not have control on that. I addressed some of your comments, and keep some unchanged to keep consistency with others.

**Pull request merged**